### PR TITLE
Add a new public method for getting the selector for a block

### PR DESCRIPTION
**Pull request by ingeniumed**
## What?
This PR came about from the conversation that was had [here](https://github.com/WordPress/gutenberg/pull/45089#discussion_r1011954379) between @oandregal, @alecgeatches  and I about having a public method to get the selector for a block instead of exposing the `get_blocks_metadata function.

## Why?
Given the [recent drive to do a code quality check for theme.json APIs](https://github.com/WordPress/gutenberg/issues/45171), it was mentioned to not expose the `get_blocks_metadata` function like we proposed in [this PR](https://github.com/WordPress/gutenberg/pull/45089). Instead, it's better to make a simpler public method that when given the block name would return the selector for that block.

This is using existing cached methods to do its job so the lookup time is minimal, and it ensure it only returns the basic information needed which is the selectors.

## How?
It's using a simpler version of the `get_blocks_metadata` function by:

- getting the registered blocks
- looking up the name of block based on the name provided
- checking to see what selectors are available, and returning one accordingly

## Testing Instructions
Make a test plugin that essentially calls the following:

```$looked_up_selector = wp_theme_get_css_selector_for_block( $block_name );```

for a block. Ensure the right value is given back.

### CODE REVIEW

**Comment by ingeniumed** [comment]
@oandregal Something that struck us while making this:

There is some overlap in the `wp_theme_get_css_selector_for_block()`, and in `get_blocks_metadata` when it comes to getting the registered blocks and their various selectors. Any way we want to reduce that overlap, or it makes sense just leaving it the way it is?

Technically, the purpose of the two functions is different as the former is for getting the selector of one block while the latter gathers all the blocks, along with some other metadata properties for each block.

**Comment by alecgeatches** [comment]
> There is some overlap in the `wp_theme_get_css_selector_for_block()`, and in `get_blocks_metadata` when it comes to getting the registered blocks and their various selectors. Any way we want to reduce that overlap, or it makes sense just leaving it the way it is?

Agree, it seems like it would make sense to implement a 6.2 version of `get_blocks_metadata()` that reuses `wp_theme_get_css_selector_for_block()` so those implementations don't need to be updated separately.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php [hunk 1]
```diff
@@ -48,6 +48,35 @@ function wp_theme_has_theme_json() {
...
 
+if ( ! function_exists( 'wp_theme_get_css_selector_for_block' ) ) {
+	/**
+	 * Lookup a CSS selector for the block provided, and return it if it exists.
+	 *
+	 * @param string $block_name The name of the block to lookup the CSS selector for.
+	 *
+	 * @return string|null the CSS selector for the block.
+	 */
+	function wp_theme_get_css_selector_for_block( $block_name ) {
```
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php:59]: The theme and the block are unrelated, so it seems better to rename this to something along the lines of `wp_block_get_css_selector`.

Note [this conversation](https://github.com/WordPress/gutenberg/issues/45194). A block has more than one selector. This method should be able to retrieve any of the selectors provided by the block.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php [hunk 2]
```diff
@@ -48,6 +48,35 @@ function wp_theme_has_theme_json() {
...
+		$registry = WP_Block_Type_Registry::get_instance();
+		$blocks   = $registry->get_all_registered();
```
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php:61]: If the method is about a single block, we can use `get_registered( $name )` instead.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php (hunk 1 above)
Comment (oandregal) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php:59]: It'd be good to update `get_blocks_metadata` to use this method as well. It'd give us feedback whether this is good in practice or not.

For example, I haven't really dug into this, but just thinking about reusing this method at `get_blocks_metadata` it struck me that an alternative to this PR could be adding a `get_css_selector` method to the `WP_Block_Type` class instead.

**Comment by oandregal** [comment]
@aaronrobertshaw given your interest and thoughts on https://github.com/WordPress/gutenberg/issues/45194 would you available to give feedback on this PR? I'm out next week, but I think we need to connect the work here with the work we talked about over there.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php (hunk 1 above)
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php:59]: > The theme and the block are unrelated, so it seems better to rename this to something along the lines of wp_block_get_css_selector.

Naive question, should we be prefixing this with `gutenberg_` prior to it landing in core?

> Note this conversation. A block has more than one selector. This method should be able to retrieve any of the selectors provided by the block.

+1 on this comment. Not only can blocks have selectors on a root and per-block-support basis, in the near future we plan on allowing custom selectors for individual styles for example on the color's background.

The point above probably illustrates we'll get even more benefit in future from reusing the results of this PR.

> an alternative to this PR could be adding a get_css_selector method to the WP_Block_Type class instead.

Sounds good to me 👍

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for working on this @ingeniumed 👍 

I think @oandregal makes a lot of great points in his feedback to date. I'll add my two cents to those comments inline.

My main thoughts are:

1. We'll definitely need to extend this to allow specification which block support (`color`, `typography`) or which individual style (`background`, `border-radius`) we want a selector for.
2. Making this a method on `WP_Block_Type` seems like a great fit to me.
3. Reusing whatever the result of this PR is in the `get_blocks_metadata` method is a win in my books.

You can find further history on the feature level selectors in https://github.com/WordPress/gutenberg/pull/42087. That PR added the ability to specify different selectors on a per block support basis. A use case for this was to allow border support styles to target the inner `img` element for an Image block instead of the outer `figure` wrapper.

**Comment by ingeniumed** [comment]
Mentioning [this PR](https://github.com/WordPress/gutenberg/pull/46496) that @aaronrobertshaw mentioned to me as it relates to what @alecgeatches and I are proposing here.

**Comment by aaronrobertshaw** [comment]
> Mentioning https://github.com/WordPress/gutenberg/pull/46496

Thanks for linking the above PR 👍 

It is a work in progress but a first pass at achieving the goals set out in https://github.com/WordPress/gutenberg/issues/45194 that was linked earlier in this PR discussion. It would eventually make the block's `selectors` something that could be accessed directly from the block-type object. Whether a util function would be handy to assist with retrieving a specific feature or style's selector, I'll leave up to others.

Unfortunately, I'll be AFK until early January. So https://github.com/WordPress/gutenberg/pull/46496 is unlikely to land before then unless someone else would like to pick that one up.

**Comment by ingeniumed** [comment]
Closing my PR given the changes that we wanted for this PR have already gone in, in #46496

**Pull request closed without merging**